File: D01/day01.py
#! /usr/bin/env python3

with open('input.txt') as f:
    data = [int(s) for s in f.read().split("\n") if s != ""]

def quantify(d, pred=bool):
    return sum(pred(dd) for dd in d)

print(quantify(zip(data,data[1:]), lambda x: x[1]>x[0]))
print(quantify(
    zip(
        zip(data,data[1:],data[2:]),
        zip(data[1:],data[2:],data[3:])
    ), lambda x: sum(x[1])>sum(x[0])))

File: day01/day01.rs
use std::fs::File;
use std::io::{self, BufReader, BufRead};
// use std::cmp::Ordering;


fn quantify<I, T>(data: &I, predicate: Option<fn(T) -> bool>) -> usize
where T: Iterator<Item=T>,
      I: Ord {
    match predicate {
        Some(p) => data.filter(p).count(),
        None => data.filter(|l| l && true).count(),
    }
}


fn main() -> io::Result<()> {
    let f = File::open("input.txt")?;
    let f = BufReader::new(f)
        .lines()
        .map(|l| l
            .unwrap()
            .parse::<u16>()
            .unwrap());  // more idiomatic way to do this other than sandwiched unwraps?

    let part1 = quantify(&f.zip(f.next()), None);
    // let part1 = quantify(f.zip(f, None);
    Ok(())
}
